jpa/subsystem/src/main/java/org/jboss/as/jpa/transaction/TransactionUtil.java

```diff
@@ -137,47 +181,50 @@ public SessionSynchronization(EntityManager session, String scopedPuName) {
         }
 
         public void beforeCompletion() {
+            afterCompletionCalled = false;
         }
 
-        public void afterCompletion(int status) {
+        public synchronized void afterCompletion(int status) {
             /**
-             * If its not safe (safeToClose returns false) to close the EntityManager now,
-             * any connections joined to the JTA transaction
-             * will be released by the JCA connection pool manager.  When the JTA Transaction is no longer
-             * referencing the EntityManager, it will be eligible for garbage collection.
-             * See AS7-6586 for more details.
+             * Note: synchronization is to protect against two concurrent threads from closing the EntityManager (manager)
```

Comment:
This is primarily the transaction timeout problem.  When the transaction times out in a remote JVM, a communication thread may call in to cancel the transaction (Synchronization.afterCompletion(int status) will run in a background thread).  The local TM Reaper may also run, it really depends on timing which occurs first.  We currently protect against the local TM reaper thread but have no way of handling the remote TM reaper case (safely).  With this change, we capture both of the conditions I mentioned below (Synchronization.afterCompletion(int) has been called and the application thread is disassociated from the TM).  Until both of these conditions are met, it is unsafe to close the EntityManager.  This fix addresses checking for both of these conditions and safely closing the EntityManager.